Refactor printing of current configuration settings

SejalBehere opened the pull request:
For Fixes #33 this PR inserts the function Ros_ConfigFile_PrintActiveConfiguration() and is used to print the current configuration settings once the parsing of the configuration file is complete

gavanderhoorn commented on src/ConfigFile.c:
```diff
@@ -678,6 +678,45 @@ void Ros_ConfigFile_ValidateNonCriticalSettings()
+void Ros_ConfigFile_PrintActiveConfiguration()
+{
+    Ros_Debug_BroadcastMsg("Config: ros_domain_id = %d", g_nodeConfigSettings.ros_domain_id);
+    Ros_Debug_BroadcastMsg("Config: node_name = %s", g_nodeConfigSettings.node_name);
```
I would suggest to enclose all strings (ie: `%s`) in single quotes (ie: `'`).

If a configuration item is left empty (or the default value is an empty string), the current implementation prints nothing (for `node_namespace` fi we'd see:

```
Config: node_namespace =
```

By enclosing the string in single quotes, the fact it's actually set to an empty string is would be much more apparent:

```
Config: node_namespace = ''
```

So for this entry specifically, you'd do:

```suggestion
    Ros_Debug_BroadcastMsg("Config: node_name = '%s'", g_nodeConfigSettings.node_name);
```

SejalBehere replied:
I will make the changes to the strings and include the namespace of all the keys.

SejalBehere replied:
I have made the changes requested to this and it is ready for another review.

gavanderhoorn reviewed (changes requested):
One in-line comment.

Additionally, I would suggest to prefix all keys with the enclosing namespace name (ie: instead of `tf`, print `publisher_qos.tf`).

gavanderhoorn commented on src/ConfigFile.c:
```diff
@@ -678,6 +678,45 @@ void Ros_ConfigFile_ValidateNonCriticalSettings()
+    Ros_Debug_BroadcastMsg("Config: stop_motion_on_disconnect = %d", g_nodeConfigSettings.stop_motion_on_disconnect);
+    Ros_Debug_BroadcastMsg("Config: inform_job_name = %s", g_nodeConfigSettings.inform_job_name);
+    Ros_Debug_BroadcastMsg("Config: allow_custom_inform_job = %d", g_nodeConfigSettings.allow_custom_inform_job);
+    Ros_Debug_BroadcastMsg("Config: userlan_monitor_enabled = %d", g_nodeConfigSettings.userlan_monitor_enabled);
+    Ros_Debug_BroadcastMsg("Config: userlan_monitor_port = %d", g_nodeConfigSettings.userlan_monitor_port);
```
Just noticed this seems to print the active configuration, as opposed to the parsed one (including defaults).

My `.yaml` has:

```yaml
userlan_monitor_enabled: true
#userlan_monitor_port: USER_LAN1
```

MotoROS2 debug listener output:

```
[2023-07-14 14:42:04.815691] [192.168.255.1:64288]: FRI 2023-07-14 13:44:51.465 Config: userlan_monitor_enabled = 1
[2023-07-14 14:42:04.824265] [192.168.255.1:64288]: FRI 2023-07-14 13:44:51.465 Config: userlan_monitor_port = 1
```

This is sort-of correct, in that after auto-detection, `userlan_monitor_port` is set to `1`. But in the `.yaml`, it's not set (leading to auto-detection being activated).

I guess I should expect this to print the active configuration, given the name of the function (`Ros_ConfigFile_PrintActiveConfiguration()`). As a debug aid though, we might be interested in the parsed configuration.

gavanderhoorn replied:
So this was more of a discussion than a request I guess.

I'm just not sure at the moment whether we'd want want over the other (ie: as-parsed or active).

What do you think?

SejalBehere replied:
I think the user might prefer to know what the active configuration is if they are going through the log for troubleshooting purposes. They will have access to the configuration file and would know what value should be shown once it is parsed; whether it's set in the .yaml or not. 

One thing we can do is comment in mentioning this that auto-detect will be activated if the value is not set in the .yaml file. That way the user will have access to the active value and we could avoid the confusion of whether the value was parsed from the .yaml or not in the case where it is not set in the .yaml.

gavanderhoorn replied:
> I think the user might prefer to know what the active configuration is if they are going through the log for troubleshooting purposes. 

Agreed.

> They will have access to the configuration file and would know what value should be shown once it is parsed; whether it's set in the .yaml or not.

well, that's the thing: do they? If the file is saved on the controller, they would not have (direct) access. Additionally, what's in the file and what the parser 'decides' is in the file are not necessarily the same thing.

> One thing we can do is comment in mentioning this that auto-detect will be activated if the value is not set in the .yaml file. That way the user will have access to the active value and we could avoid the confusion of whether the value was parsed from the .yaml or not in the case where it is not set in the .yaml.

I'm not sure I understand this part of your comment completely.

SejalBehere replied:
> well, that's the thing: do they? If the file is saved on the controller, they would not have (direct) access.

I get that. Would it be excessive if we print out both the active and the parsed value? Seeing how both of them would be important for the user, depending upon whether the value is set in the .yaml or not and if it is, then what value is being parsed, it might be better to display both of the values. 

> I'm not sure I understand this part of your comment completely.

I meant to say that we can mention in the .yaml that auto-detect would be activated if there's no value set in the config file but all of that is already mentioned in the comments in the .yaml. Apologies for the oversight.

ted-miller replied:
Personally, I think printing both values would be excessive. I think printing the "active" settings is best, regardless of whether those settings were read from the config file or calculated through the code.

gavanderhoorn replied:
The active configuration would indeed be the one that has the most influence on the runtime behaviour of MotoROS2. In that sense it would be the one to log -- anything not as it should be in there would be reason to start debugging.

I believe I'd agree with @ted-miller in that printing both might be too much.

Let's just keep the implementation as-is, but I'll make one small suggestion that would make things somewhat clearer.

gavanderhoorn replied:
Thanks for the input @SejalBehere and @ted-miller.

gavanderhoorn commented on src/ConfigFile.c:
```diff
@@ -753,46 +792,11 @@ void Ros_ConfigFile_Parse()
         }
 
-    } while (!bOkToInit);
+    } while (!bOkToInit); 
```
```suggestion
    } while (!bOkToInit);
```

gavanderhoorn reviewed (approved):
This looks better now.

I can't test this, but if you could post a copy of the debug log lines this prints at startup just to show it's correct, we can merge.

I'd also recommend removing the trailing whitespace that was introduced.

gavanderhoorn commented on src/ConfigFile.c:
```diff
@@ -678,6 +678,45 @@ void Ros_ConfigFile_ValidateNonCriticalSettings()
+    Ros_Debug_BroadcastMsg("Config: update_periods.executor_sleep_period = %d", g_nodeConfigSettings.executor_sleep_period);
+    Ros_Debug_BroadcastMsg("Config: update_periods.action_feedback_publisher_period = %d", g_nodeConfigSettings.action_feedback_publisher_period);
+    Ros_Debug_BroadcastMsg("Config: update_periods.controller_status_monitor_period = %d", g_nodeConfigSettings.controller_status_monitor_period);
+    Ros_Debug_BroadcastMsg("Config: publisher_qos.robot_status = %d", g_nodeConfigSettings.qos_robot_status);
+    Ros_Debug_BroadcastMsg("Config: publisher_qos.joint_states = %d", g_nodeConfigSettings.qos_joint_states);
+    Ros_Debug_BroadcastMsg("Config: publisher_qos.tf = %d", g_nodeConfigSettings.qos_tf);
```
minor, but these lines will now only print a nr (ie: the value of the `enum` member). They used to print the name of the QoS profile.

It would be nice if we could add a mapping from the `enum` integer to their string description again, as that would be much more human readable, but I won't block merging this PR for that (seeing as this was already accepted as part of #81.

gavanderhoorn commented on src/ConfigFile.c:
```diff
@@ -773,45 +812,10 @@ void Ros_ConfigFile_Parse()
 
     } while (!bOkToInit);
-
+    
```
```suggestion

```

gavanderhoorn commented:
@SejalBehere wrote:
> Also, I am looking for a way to add a mapping from the enum integer to their string description.

I would perhaps suggest a simple `switch(..)`, like we do here:

https://github.com/Yaskawa-Global/motoros2/blob/ab02d18c96fe6b388eb10b7f9148bff01de8a9b4/src/ErrorHandling.c#L11-L33

you'd have fewer values to check (there are only a couple of values in [this enum](https://github.com/Yaskawa-Global/motoros2/blob/ab02d18c96fe6b388eb10b7f9148bff01de8a9b4/src/ConfigFile.h#L66-L75)), but the idea could be the same.

I might suggest not using `strncpy(..)` though. Returning `static char[] ...` might be less involved in this case (@ted-miller should be able to show you the pattern).

There *could* be something in either RCL or RCLC that already implements this (ie: `enum` value to `char[]` for the Qos values). That might save you some work, but I wouldn't be able to point you to it.

SejalBehere commented:
Thank you @gavanderhoorn for the guidance. I'm working on this and will put up the modified code for review.

SejalBehere commented:
I have updated ConfigFile.c to add a lookup table for mapping from integers to strings for `publisher_qos` namespace. I have also refactored `Ros_ConfigFile_CheckYamlEvent` and used the same lookup table instead of the if/else hardcoded string. This PR is ready for another review.

I tested this on the controller and the output log displayed is as follows:

[1689882942.78701329] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.312 Config: ros_domain_id = 0 
[1689882942.78710938] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.312 Config: node_name = 'motoman'
[1689882942.78720188] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.312 Config: node_namespace = ''
[1689882942.78730178] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.312 Config: remap_rules = ''
[1689882942.78738475] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.312 Config: agent_ip_address = '192.168.1.50'
[1689882942.78741455] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.312 Config: agent_port_number = '8888'
[1689882942.78746367] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.312 Config: sync_timeclock_with_agent = 1
[1689882942.78754473] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.313 Config: namespace_tf = 1
[1689882942.78761578] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.313 Config: publish_tf = 1
[1689882942.78767252] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.313 List of configured joint names:
[1689882942.78775954] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.313 ---
[1689882942.78783751] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.313 x
[1689882942.78788447] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.313 x
[1689882942.78791237] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.313 x
[1689882942.78794026] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.313 x
[1689882942.78796816] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.313 x
[1689882942.78799605] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.313 x
[1689882942.78804374] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.313 x
[1689882942.78808808] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.313 x
[1689882942.78811598] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.313 ---
[1689882942.78816962] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.313 x
[1689882942.78819799] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.313 x
[1689882942.78822613] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.313 x
[1689882942.78825402] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.313 x
[1689882942.78829455] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.313 x
[1689882942.78835988] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.313 x
[1689882942.78838897] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.313 x
[1689882942.78844595] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.313 x
[1689882942.78847361] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.313 ---
[1689882942.78852773] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.313 x
[1689882942.78855562] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.313 x
[1689882942.78858328] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.313 x
[1689882942.78861070] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.314 x
[1689882942.78867054] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.314 x
[1689882942.78871441] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.314 x
[1689882942.78875566] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.314 x
[1689882942.78881073] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.314 x
[1689882942.78883862] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.314 ---
[1689882942.79081631] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.314 x
[1689882942.79106975] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.316 x
[1689882942.79112935] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.316 x
[1689882942.79119349] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.316 x
[1689882942.79127884] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.317 x
[1689882942.79133821] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.317 x
[1689882942.79136801] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.317 x
[1689882942.79143405] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.317 x
[1689882942.79148269] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.317 ---
[1689882942.79152489] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.317 x
[1689882942.79158998] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.317 x
[1689882942.79163289] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.317 x
[1689882942.79167581] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.317 x
[1689882942.79171872] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.317 x
[1689882942.79174709] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.317 x
[1689882942.79180717] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.317 x
[1689882942.79183578] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.317 x
[1689882942.79187775] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.317 ---
[1689882942.79193926] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.317 x
[1689882942.79198217] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.317 x
[1689882942.79201078] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.317 x
[1689882942.79205251] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.317 x
[1689882942.79209399] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.317 x
[1689882942.79213572] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.317 x
[1689882942.79216433] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.317 x
[1689882942.79219246] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.317 x
[1689882942.79223967] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.317 ---
[1689882942.79229856] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.317 x
[1689882942.79234076] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.317 x
[1689882942.79238224] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.318 x
[1689882942.79242396] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.318 x
[1689882942.79245210] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.318 x
[1689882942.79263949] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.318 x
[1689882942.79268360] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.318 x
[1689882942.79273057] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.318 x
[1689882942.79276061] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.318 ---
[1689882942.79281092] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.318 x
[1689882942.79286242] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.318 x
[1689882942.79290748] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.318 x
[1689882942.79293609] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.318 x
[1689882942.79296470] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.318 x
[1689882942.79299283] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.318 x
[1689882942.79303336] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.318 x
[1689882942.79306197] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.318 x
[1689882942.79308987] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.318 ---
[1689882942.79315782] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.319 Config: logging.log_to_stdout = 0
[1689882942.79321742] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.319 Config: update_periods.executor_sleep_period = 10
[1689882942.79326105] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.319 Config: update_periods.action_feedback_publisher_period = 20
[1689882942.79328990] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.319 Config: update_periods.controller_status_monitor_period = 10
[1689882942.79331851] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.319 Config: publisher_qos.robot_status = 'sensor_data'
[1689882942.79334688] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.319 Config: publisher_qos.joint_states = 'sensor_data'
[1689882942.79337502] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.319 Config: publisher_qos.tf = 'default'
[1689882942.79341698] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.319 Config: tf_frame_prefix = ''
[1689882942.79359031] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.319 Config: stop_motion_on_disconnect = 1
[1689882942.79366231] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.319 Config: inform_job_name = 'INIT_ROS'
[1689882942.79369569] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.319 Config: allow_custom_inform_job = 0
[1689882942.79372621] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.319 Config: userlan_monitor_enabled = 1
[1689882942.79375529] [192.168.1.31:59091]: THU 2023-07-20 15:55:43.319 Config: userlan_monitor_port = 1

ted-miller commented:
This looks ok to me. I'll defer to @gavanderhoorn though since it was his initial request. (At least I think it was. My memory is fuzzy.)

gavanderhoorn commented:
Closing in favour of #127.

Thanks for all the initial work @SejalBehere 👍
